sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/SymbolUsageCollector.cs

```diff
@@ -0,0 +1,284 @@
+
+        private IEnumerable<ISymbol> GetSymbols<TSyntaxNode>(TSyntaxNode node, Func<TSyntaxNode, bool> condition)
+            where TSyntaxNode : SyntaxNode
+        {
+            return condition(node)
+                ? GetCandidateSymbols(getSemanticModel(node).GetSymbolInfo(node))
+                : Enumerable.Empty<ISymbol>();
+
+            IEnumerable<ISymbol> GetCandidateSymbols(SymbolInfo symbolInfo)
+            {
+                return new[] { symbolInfo.Symbol }
+                    .Concat(symbolInfo.CandidateSymbols)
+                    .Select(GetOriginalDefinition)
+                    .WhereNotNull();
+
+                ISymbol GetOriginalDefinition(ISymbol candidateSymbol)
+                {
+                    if (candidateSymbol is IMethodSymbol methodSymbol &&
+                        methodSymbol.MethodKind == MethodKind.ReducedExtension)
+                    {
+                        return methodSymbol.ReducedFrom?.OriginalDefinition;
+                    }
+                    return candidateSymbol?.OriginalDefinition;
+                }
+            }
+        }
+
+        private void TryStorePropertyAccess(ExpressionSyntax node, List<ISymbol> identifierSymbols)
+        {
+            var propertySymbols = identifierSymbols.OfType<IPropertySymbol>();
+            if (propertySymbols.Any())
+            {
+                var access = EvaluatePropertyAccesses(node);
+                foreach (var propertySymbol in propertySymbols)
+                {
+                    StorePropertyAccess(propertySymbol, access);
+                }
+            }
+        }
+
+        private void StorePropertyAccess(IPropertySymbol propertySymbol, AccessorAccess access)
+        {
+            if (PropertyAccess.ContainsKey(propertySymbol))
+            {
+                PropertyAccess[propertySymbol] |= access;
+            }
+            else
+            {
+                PropertyAccess[propertySymbol] = access;
+            }
+        }
+
+        private AccessorAccess EvaluatePropertyAccesses(ExpressionSyntax node)
+        {
+            var topmostSyntax = GetTopmostSyntaxWithTheSameSymbol(node);
+
+            if (topmostSyntax.Parent is AssignmentExpressionSyntax assignmentExpression)
+            {
+                if (assignmentExpression.IsKind(SyntaxKind.SimpleAssignmentExpression))
+                {
+                    // Prop = value --> set
+                    // value = Prop --> get
+                    return assignmentExpression.Left == topmostSyntax
+                        ? AccessorAccess.Set
+                        : AccessorAccess.Get;
+                }
+                else
+                {
+                    // Prop += value --> get/set
+                    return AccessorAccess.Both;
+                }
+            }
+
+            // nameof(Prop) --> get/set
+            if (node.IsInNameofCall(getSemanticModel(node)))
+            {
+                return AccessorAccess.Both;
```

Comment:
Why is nameof(X) treated as accessing both get and set?

Reply:
Which one should it be? :)

If we assume that when a property should not be removed when it is used in a `nameof` expression, I think we should mark both accessors as used. On the other hand, perhaps we should suggest members for removal if they are only used in `nameof`... I didn't check if the ITs will raise many issues when the `nameof` exclusion is removed.

This functionality was added two years ago when the property support was implemented and I decided to keep it like this.